### PR TITLE
Fix foundry-remove-node failing to remove terminal nodes

### PR DESCRIPTION
**Pull request by nwatson22**
This PR fixes a bug where when you used `foundry-remove-node` to remove a node that pruned off a terminal node, the node's ID would remain in the terminal nodes list, and if dereferenced later, would cause the proof to crash.

### CODE REVIEW

**Comment by ehildenb** [comment]
This seems like something that should be fixed in `APRProof`, by adding an `APRProof.prune_from(node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]`

It should be implemented something like:

```
def prune_from(node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]:
    pruned_nodes = self.kcfg.prune(node_id, keep_nodes=list(keep_nodes) + [self.target, self.init])
    for nid in pruned_nodes:
        self._terminal_nodes.discard(nid)
    return pruned_nodes
```

Then `APRBMCProof.prune_from` should do:

```
def prune_from(node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]:
    pruned_nodes = super().prune_from(node_id, keep_nodes=keep_nodes)
    for nid in pruned_nodes:
        self._bounded_nodes.discard(nid)
    return pruned_nodes
```

**Comment by nwatson22** [comment]
Currently blocked on this by #1969 . When the pyk version is updated we will switch to using the new functions from https://github.com/runtimeverification/pyk/pull/561.

**Comment by nwatson22** [comment]
> 

Added in pyk https://github.com/runtimeverification/pyk/pull/561 and now using these methods here.

**Comment by nwatson22** [comment]
> This seems like something that should be fixed in `APRProof`, by adding an `APRProof.prune_from(node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]`
> 
> It should be implemented something like:
> 
> ```
> def prune_from(node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]:
>     pruned_nodes = self.kcfg.prune(node_id, keep_nodes=list(keep_nodes) + [self.target, self.init])
>     for nid in pruned_nodes:
>         self._terminal_nodes.discard(nid)
>     return pruned_nodes
> ```
> 
> Then `APRBMCProof.prune_from` should do:
> 
> ```
> def prune_from(node_id: NodeIdLike, keep_nodes: Iterable[NodeIdLike]) -> list[NodeIdLike]:
>     pruned_nodes = super().prune_from(node_id, keep_nodes=keep_nodes)
>     for nid in pruned_nodes:
>         self._bounded_nodes.discard(nid)
>     return pruned_nodes
> ```

Added in pyk https://github.com/runtimeverification/pyk/pull/561 and now using these methods here.

File: kevm-pyk/src/kevm_pyk/foundry.py [hunk 1]
```diff
@@ -690,7 +690,9 @@ def foundry_remove_node(foundry_root: Path, test: str, node: NodeIdLike) -> None
     contract_name, test_name = test.split('.')
     proof_digest = foundry.proof_digest(contract_name, test_name)
     apr_proof = APRProof.read_proof_data(apr_proofs_dir, proof_digest)
-    node_ids = apr_proof.kcfg.prune(node, [apr_proof.init, apr_proof.target])
+
+    node_ids = apr_proof.prune_from(node, [apr_proof.init, apr_proof.target])
+
```
Comment (nwatson22) [inline kevm-pyk/src/kevm_pyk/foundry.py:695]:
```suggestion
    node_ids = apr_proof.prune_from(node, [apr_proof.init, apr_proof.target])
```

File: kevm-pyk/src/tests/integration/test_foundry_prove.py [hunk 2]
```diff
@@ -206,6 +206,32 @@ def test_foundry_auto_abstraction(foundry_root: Path, update_expected_output: bo
...
+        foundry_root=foundry_root,
+        test='AssumeTest.test_assume_true',
+        node=4,
+    )
+    prove_res = foundry_prove(
+        foundry_root,
+        tests=['AssumeTest.test_assume_true'],
+        max_depth=0,
+    )
+    assert_fail('AssumeTest.test_assume_true', prove_res)
```
Comment (ehildenb) [inline kevm-pyk/src/tests/integration/test_foundry_prove.py:226]: Why do we need the extra call to `foundry_prove` here? Can't we just assert that the proof is now pending?

**Comment by ehildenb** [comment]
I believe also `kevm prune-proof` needs to be updated.

File: kevm-pyk/src/tests/integration/test_foundry_prove.py (hunk 2 above)
Comment (nwatson22) [reply kevm-pyk/src/tests/integration/test_foundry_prove.py:226]: The foundry_x level functions don't actually return the proof. This was the only way I see to get the proof result. I will refactor this in a future PR.

File: kevm-pyk/src/kevm_pyk/foundry.py [hunk 3]
```diff
@@ -690,7 +690,7 @@ def foundry_remove_node(foundry_root: Path, test: str, node: NodeIdLike) -> None
     contract_name, test_name = test.split('.')
     proof_digest = foundry.proof_digest(contract_name, test_name)
     apr_proof = APRProof.read_proof_data(apr_proofs_dir, proof_digest)
-    node_ids = apr_proof.kcfg.prune(node, [apr_proof.init, apr_proof.target])
+    node_ids = apr_proof.prune_from(node, [apr_proof.init, apr_proof.target])
```
Comment (nwatson22) [inline kevm-pyk/src/kevm_pyk/foundry.py:693]:
```suggestion
    node_ids = apr_proof.prune_from(node)
```
Redundant

**Comment by ehildenb** [comment]
It looks like we need to make this argument optional: https://github.com/runtimeverification/pyk/blob/376b83f6499e1b38b9e5a48ef7dc938992364bc0/src/pyk/proof/reachability.py#L140

**Comment by nwatson22** [comment]
> It looks like we need to make this argument optional: https://github.com/runtimeverification/pyk/blob/376b83f6499e1b38b9e5a48ef7dc938992364bc0/src/pyk/proof/reachability.py#L140
> 
> ```
> keep_nodes: Iterable[NodeIdLike] = ()
> ```

Made optional here: https://github.com/runtimeverification/pyk/pull/591

**Comment by nwatson22** [comment]
> I believe also `kevm prune-proof` needs to be updated.

Updated and tested node pruning locally with non-foundry proof.

**Review by ehildenb — APPROVED** [review]

**Pull request merged**